### PR TITLE
[SPARK-45967][BUILD] Upgrade jackson to 2.16.0

### PR DESCRIPTION
**Pull request by panbingkun**
### What changes were proposed in this pull request?
The pr aims to upgrade FasterXML jackson from 2.15.2 to 2.16.0.

### Why are the changes needed?
New version that fix some bugs, release notes as follows:
- 2.1.6.0 https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.16, eg:
[Databind](https://github.com/FasterXML/jackson-databind) [#1770](https://github.com/FasterXML/jackson-databind/issues/1770): Incorrect deserialization for BigDecimal numbers
- 2.15.3 https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.15.3, eg:
[Databind](https://github.com/FasterXML/jackson-databind) [#3968](https://github.com/FasterXML/jackson-databind/issues/3968): Records with additional constructors failed to deserialize

The last upgrade occurred 6 months ago, https://github.com/apache/spark/pull/41414

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

### CODE REVIEW

**Comment by LuciferYang** [comment]
cc @pjfanning ~ Are there any areas to focus on when upgrading to use Jackson 2.16.0? Thanks ~

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 1]
```diff
@@ -1175,7 +1175,8 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
     val invalidDataType = "MAP<INT, cow>"
     val invalidDataTypeReason = "Unrecognized token 'MAP': " +
       "was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n " +
-      "at [Source: (String)\"MAP<INT, cow>\"; line: 1, column: 4]"
+      "at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); " +
+      "line: 1, column: 4]"
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:1179]: image

File: sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala [hunk 2]
```diff
@@ -1190,7 +1191,8 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
     val invalidTableSchema = "x INT, a cow"
     val invalidTableSchemaReason = "Unrecognized token 'x': " +
       "was expecting (JSON String, Number, Array, Object or token 'null', 'true' or 'false')\n" +
-      " at [Source: (String)\"x INT, a cow\"; line: 1, column: 2]"
+      " at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); " +
+      "line: 1, column: 2]"
```
Comment (panbingkun) [inline sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:1195]: dito

**Comment by pjfanning** [comment]
@LuciferYang it appears you are hitting some issues with tests but other than that, I don't expect significant issues. The changes are less significant than in 2.15.x.

**Review by dongjoon-hyun — COMMENTED** [review]
This looks fine if CI passes.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Merged to master.

**Comment by LuciferYang** [comment]
> https://issues.apache.org/jira/browse/SPARK-42854

Thanks for your explain @pjfanning

**Comment by LuciferYang** [comment]
late LGTM

**Pull request closed without merging**